src/model/mlp.py

```diff
@@ -136,26 +133,23 @@ def _compute_error(self, target):
             a numpy array (1,nOut) containing the output of the layer
         """
         output = self._get_output_layer().outp
-        return self.loss.calculateError(target, output)
+        return self.loss.calculateError(self._get_encoded_label(target), output)
     
     def _update_weights(self, learningRate, label):
         """
         Update the weights of the layers by propagating back the error
         """
-
         tempWeights = None
         tempDerivatives = None
         for layer in reversed(self.layers):
             if layer == self._get_output_layer():
-
                 layer.computeDerivative(self.loss.calculateDerivative(label, layer.outp), 1.0)
                 tempWeights = layer.weights
                 tempDerivatives = layer.deltas
             else:
-                layer.computeDerivative(tempDerivatives, tempWeights)
+                layer.computeDerivative(tempDerivatives, tempWeights[1:])
```

Comment:
Ist das, damit der Bias bei der Ableitung nicht berücksichtigt wird? Falls ja: Wieso sollte das so sein? Dazu finde ich nichts in den Folien.

Reply:
Ja. Wir haben eine Ableitung pro Neuron in der nächsten Schichte. Für die aktuelle Schichte, für jedes Neuron, haben wir den Fehler aller Neuronen der nächsten Schichte, die sich mit diesem Neuron der aktuellen Schichte verbinden. Das ist das "Downstream" von der Folien. 

Mit Zahlen: 

Hidden Layer: 128 Neuronen
Output Layer: 10 Neuronen

das entspricht: 
- 129x10 Gewichte in Output Layer (jedes Neuron, 128 Verbindungen mit der Hidden Layer + Bias)
- 10x1 Deltas in Output Layer(1 Delta pro Neuron) 
- Deltas der Hidden Layer müssen 128 Elemente haben (1 Delta pro Neuron). D.h: deltas_hidden = weights_output (ohne Bias, da der Bias keine Verbindung zwischen Schichten ist) * deltas_output

Reply:
Klar kann man beim Bias nicht weiter propagieren, aber er muss doch auch aktualisiert werden.

Reply:
Er wird auch aktualisiert. Bei dem Update wird er aktualisiert da er auch ein Input des Neurons ist. Der Bias hat nur keinen Einfluss auf die Deltas.

![image](https://user-images.githubusercontent.com/7738765/41819596-3626d350-77c3-11e8-8e7f-48da67d753b5.png)

Reply:
Weitere Anmerkungen? Sonst würde ich sagen dass wir die Änderungen mergen und mit dem Bericht anfangen.